### PR TITLE
Site Settings: Request site post types upon content types toggling

### PR DESCRIPTION
**Pull request by tyxla**
Currently, enabling/disabling a post type in the Custom Content Types card does not update the sidebar with the latest post types. For example, if the **Testimonial** post type is enabled, and you disable it, and it's not supported by the active theme, it should be removed from the sidebar right away. Right now, in order to see it removed, you need to refresh Calypso. 

This PR fixes this issue by:
* Introducing support for callbacks to `handleAutosavingToggle` in the `wrapSettingsForm` HoC.
* Requesting the site post types by passing the request as a callback after settings are saved.

This PR is part of #13568.

To test:
* Checkout this branch.
* Go to `/settings/writing/$site`, where `$site` is one of your Jetpack sites (wp.com sites will be supported too when we land #14572).
* Activate a theme that does not support the **Testimonial** post type.
* Activate the **Testimonial** post type and verify sure the **Testimonial** item shows in the sidebar menu shortly.
* Deactivate the **Testimonial** post type and verify the **Testimonial** item gets removed from the sidebar menu shortly.

### CODE REVIEW

File: client/my-sites/site-settings/custom-content-types/index.jsx [hunk 1]
```diff
@@ -52,6 +53,13 @@ class CustomContentTypes extends Component {
 		return isRequestingSettings || isSavingSettings;
 	}
 
+	requestSitePostTypes = () => {
+		const { siteId } = this.props;
+		if ( siteId ) {
+			this.props.requestPostTypes( siteId );
```
Comment (seear) [inline client/my-sites/site-settings/custom-content-types/index.jsx]: I'm having trouble finding whereabouts `requestPostTypes` is bound to `dispatch`?

Comment (tyxla) [inline client/my-sites/site-settings/custom-content-types/index.jsx]: See https://github.com/Automattic/wp-calypso/pull/14576/files#diff-5832a88e66f1c149f5666b93306f55e3R176 - since we're passing an object to `mapDispatchToProps`, actions are bound automatically to `dispatch`.

Comment (seear) [inline client/my-sites/site-settings/custom-content-types/index.jsx]: gotcha 👍

**Comment by seear** [comment]
Would using a query component gated by, for example, last save time, fit better with Calypso data architecture than using the callback to fetch the post types?

**Comment by tyxla** [comment]
> Would using a query component gated by, for example, last save time, fit better with Calypso data architecture than using the callback to fetch the post types?

Sounds like an interesting solution that would probably work well, too. And it could be a better fit, but I'm not really sure - throughout Calypso callbacks are used much more, which is why I went with callbacks.

**Comment by seear** [comment]
> throughout Calypso callbacks are used much more

Interesting. I haven't seen component-level callbacks for data fetching used for a while. I'm more used to seeing any action chaining happening inside the actions themselves, though in this case I can appreciate that we don't want to fetch the post types after every settings save.

**Comment by tyxla** [comment]
I'd be happy to see another opinion before we land it anyway - @youknowriad worked a lot on the `wrapSettingsForm` HoC, what are your thoughts about using a callback for that case? Let me know if you need additional context/information.

File: client/my-sites/site-settings/wrap-settings-form.jsx [hunk 2]
```diff
@@ -108,14 +108,20 @@ const wrapSettingsForm = getFormSettings => SettingsForm => {
...
+			const settingsRequests = [
+				this.props.saveSiteSettings( siteId, pick( fields, settingsFields.site ) )
+			];
+
 			if ( jetpackSettingsUISupported ) {
-				this.props.updateSettings( siteId, pick( fields, settingsFields.jetpack ) );
+				settingsRequests.push( this.props.updateSettings( siteId, pick( fields, settingsFields.jetpack ) ) );
 			}
+
+			Promise.all( settingsRequests )
```
Comment (oskosk) [inline client/my-sites/site-settings/wrap-settings-form.jsx]: What about returning this promise instead of using a callback pattern?

The callback pattern is used on some methods here because `setState()` takes a callback arg, due to an async behaviour when updating the component's state but this is not the case here. 

Also take in account that this callback or the resolution of the Promise would be executed even when the request fails. This is not a problem here, because an extra request on failure wouldn't do any harm.

**Comment by youknowriad** [comment]
@tyxla I've not looked in details to see if feasible or not but the first thought that comes to my mind is that it shouldn't be the concern of the `HoC` to trigger the postTypes request again if the settings value change, but maybe the Component showing the sidebar CPT link should use a selector to retrieve this setting from the store and refetch if necessary?

File: client/my-sites/site-settings/wrap-settings-form.jsx (hunk 2 above)
Comment (tyxla) [inline client/my-sites/site-settings/wrap-settings-form.jsx]: Nice suggestion, it could work. I'll play with it tomorrow. Thanks!

**Comment by tyxla** [comment]
@youknowriad, the sidebar is doing exactly that. The problem is that the post types need to be refetched after the custom content type settings change. So it's about triggering a post types refetch after a successful settings save that includes one of the settings that manage the content types. @oskosk has suggested a good way forward that's worth playing with.

**Comment by youknowriad** [comment]
@tyxla can't we retrieve this "custom content type settings" with a selector in the sidebar? And trigger the `fetch` if the value of this setting changes? 

I'm assuming the successful settings save will reflect these changes in the store.

File: client/my-sites/site-settings/wrap-settings-form.jsx (hunk 2 above)
Comment (tyxla) [inline client/my-sites/site-settings/wrap-settings-form.jsx]: @oskosk actually, this will require a major refactor of `lib/track-form` - since as you mentioned, callbacks are used for `setState()`, but this way we lose the opportunity to return the promise (see `handleAutosavingRadio` for example). Maybe we could leave that for another PR, what do you think?

**Comment by tyxla** [comment]
> @tyxla can't we retrieve this "custom content type settings" with a selector in the sidebar? And trigger the fetch if the value of this setting changes?

Yes we can, but the actual setting value does no good to us in the sidebar. The setting only controls whether a post type is enabled or not, while the post types endpoint returns a lot more information about the post types, and that information is required in order to display the menu items properly.

**Comment by youknowriad** [comment]
> The setting only controls whether a post type is enabled or not.

That's a good reason to monitor changes to this setting in the element showing it (the sidebar menu item). I'm afraid the solution proposed in this PR doesn't scale enough. I mean it only monitors settings changes using the HoC. One could trigger the same settings save request in another component and this won't update the sidebar. 

Also, this kind of breaks the unidirectional data flow.

**Comment by tyxla** [comment]
@youknowriad thanks for taking the time to dissuade me from going with this approach. You're right it's too settings-specific, and we might want to use it for another section, too.

In order to avoid breaking the unidirectional data flow, I'm going to play with a solution directly in the `QueryPostTypes` query component. 

So, closing this one. Thanks for the feedback, everybody 💯

**Comment by tyxla** [comment]
I've suggested an alternative, much cleaner approach in #14621, which directly implements the logic in the `QueryPostTypes` query component. I'd appreciate some 👀  on it, thanks a ton!
